hoomd/pytest/test_box_resize.py

```diff
@@ -0,0 +1,205 @@
+            assert variant(timestep) == box_resize.variant(timestep)
+
+    def test_get_box(self, device, simulation_factory,
+                     get_snapshot, sys, box_resize):
+        sys1, make_sys_halfway, sys2 = sys
+        sys_halfway = make_sys_halfway(self._power)
+
+        sim = hoomd.Simulation(device)
+        sim.create_state_from_snapshot(get_snapshot())
+
+        sim.operations.updaters.append(box_resize)
+        sim.run(_t_start + _t_ramp)
+
+        assert box_resize.get_box(0) == sys1[0]
+        assert box_resize.get_box(_t_mid) == sys_halfway[0]
+        assert box_resize.get_box(_t_start + _t_ramp) == sys2[0]
+
+    def test_update(self, device, simulation_factory,
+                    get_snapshot, sys, box_resize):
+        sys1, _, sys2 = sys
+
+        sim = hoomd.Simulation(device)
+        sim.create_state_from_snapshot(get_snapshot())
+        sim.operations.updaters.append(box_resize)
+
+        box_resize.update(sim.state, sys2[0])
+
+        assert sim.state.box == sys2[0]
+        npt.assert_allclose(
+            sim.state.snapshot.particles.position, sys2[1])
+
+    def test_position_scale(self, device, simulation_factory,
+                            get_snapshot, sys, box_resize):
+        sys1, make_sys_halfway, sys2 = sys
+        sys_halfway = make_sys_halfway(self._power)
+
+        sim = hoomd.Simulation(device)
+        sim.create_state_from_snapshot(get_snapshot())
+        sim.operations.updaters.append(box_resize)
+
+        # Run up to halfway point
+        sim.run(_t_mid + 1)
+        assert sim.state.box == sys_halfway[0]
+        npt.assert_allclose(
+            sim.state.snapshot.particles.position, sys_halfway[1])
+
+        # Finish run
+        sim.run(_t_mid)
+        assert sim.state.box == sys2[0]
+        npt.assert_allclose(
+            sim.state.snapshot.particles.position, sys2[1])
+
+    def test_no_position_scale(self, device, simulation_factory,
+                               get_snapshot, sys):
+        sys1, make_sys_halfway, sys2 = sys
+        sys_halfway = make_sys_halfway(self._power)
+
+        variant = hoomd.variant.Power(0., 1., self._power, _t_start, _t_ramp)
+        trigger = hoomd.trigger.After(variant.t_start)
+        box_resize = hoomd.update.BoxResize(
+            box1=sys1[0], box2=sys2[0],
+            variant=variant, trigger=trigger, scale_particles=False
+        )
+
+        sim = hoomd.Simulation(device)
+        sim.create_state_from_snapshot(get_snapshot())
+        sim.operations.updaters.append(box_resize)
+
+        # Run up to halfway point
+        sim.run(_t_mid + 1)
+        assert sim.state.box == sys_halfway[0]
+        npt.assert_allclose(sim.state.snapshot.particles.position, sys1[1])
+
+        # Finish run
+        sim.run(_t_mid)
+        assert sim.state.box == sys2[0]
+        npt.assert_allclose(sim.state.snapshot.particles.position, sys1[1])
+
+
+class TestLinearVolume(TestBoxResize):
```

Comment:
I think the idea here would be to test the factory method `hoomd.update.BoxResize.linear_volume()`. Basically we just want to test that the volume changes linearly over the box change.

Reply:
I tried to test if box volume changed linearly with code like this:
```python
    def test_linear_volume(self, device, simulation_factory
        sim = hoomd.Simulation(device)
        sim.create_state_from_snapshot(get_snapshot())
        sim.operations.updaters.append(box_resize)
        sim.run(0)

        slope = (sys2[0].volume - sys1[0].volume)/_t_ramp
        for ts in range(_t_ramp + 1):
            estimated_volume = sys1[0].volume + ts*slope
            assert box_resize.get_box(ts + _t_start).volume == estimated_volume
```
and the test failed.

When I plot box volume for a 3D box, I get this:
![linear_volume_3d](https://user-images.githubusercontent.com/47036428/98593160-c16a1a00-22a0-11eb-9356-d4f7d9c2695d.png)



and when I plot box area for a 2D, I get this:
![linear_volume2d](https://user-images.githubusercontent.com/47036428/98593148-bd3dfc80-22a0-11eb-8b3a-cbd9209e385e.png)



Am I testing/interpreting this correctly?

Reply:
Does that mean we don't need to test things like `get_box` and `trigger` for the factory method? And what does the `variant` control? If I understand correctly, in order for `hoomd.update.BoxResize.linear_volume` is supposed to scale volume linearly, it uses it's own calculations to determine the appropriate coefficients for the `hoomd.variant.Power`.

Reply:
@klywang Are you test boxes cubic?

Reply:
@joaander These are not.

Reply:
Does the plot look right for cubic boxes? I'm thinking that the the power variant math @b-butler wrote may have am implicit assumption that the box is cubic.

Reply:
I think it does. I didn't realize it when writing it though.

Reply:
To get linear volume change we need to either change the way `BoxResize` works or have an internal flag that changes the behavior to allow for linear volume resizing.

Reply:
Here are the results for `BoxResize.linear_volume` for 3D cubes:
![box_resize3D](https://user-images.githubusercontent.com/47036428/99443452-d5d59480-28e8-11eb-964b-d968fdef1376.png)

And 2D squares:
![box_resize2D](https://user-images.githubusercontent.com/47036428/99443430-d0784a00-28e8-11eb-92b7-1099d1694a72.png)